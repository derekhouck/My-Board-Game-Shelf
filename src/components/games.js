import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { fetchGames } from '../actions/games';
import GamesSearchFrom from './games-search-form';

export class Games extends React.Component {
  componentDidMount() {
    this.props.dispatch(fetchGames());
  }

  renderGames() {
    let body;

    if (this.props.error) {
      body = (
        <div className="message message-error">{this.props.error.message}</div>
      );
    } else if (this.props.loading) {
      body = (
        <div className="message message-default">Loading users...</div>
      );
    } else {
      const games = this.props.games.map(game => (
        <li className="game" key={game.id} id={game.id}>
          <h3 className="game__title">{game.title}</h3>
          <ul>
            <li>Players: {game.players.min} - {game.players.max}</li>
            <li><Link to={`/games/${game.id}/edit`}>Edit</Link></li>
          </ul>
        </li>
      ));
      body = (
        <section className="games">
          <GamesSearchFrom />
          <ul className="games__list">
            {games}
          </ul>
        </section>
      );
    }
    return body;
  }

  render() {
    return (
      <section>
        <h2>Your Games</h2>
        <Link to="/games/add">Add a game</Link>
        {this.renderGames()}
      </section>
    );
  }
}

const mapStateToProps = state => ({
  games: state.games.games,
  loading: state.games.loading,
  error: state.games.error
});

export default connect(mapStateToProps)(Games);