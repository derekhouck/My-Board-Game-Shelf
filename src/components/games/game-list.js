import React from 'react';
import Game from './game';

export default function GameList(props) {
  const { controls, games } = props;

  return (
    <div className="games">
      {games.length === 0
        ? <div className="games__no-games">
          <p>There are no games to list.</p>
        </div>
        : <ul className="games__list">
          {games.map(game =>
            <Game
              controls={controls}
              game={game}
              key={game.id}
              removeButton={!!game.isUserGame}
            />
          )}
        </ul>}
    </div>
  );
}

GameList.defaultProps = {
  controls: false,
  games: []
};