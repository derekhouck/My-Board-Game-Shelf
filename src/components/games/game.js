import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { removeGameFromShelf, addGameToShelf } from '../../actions/users';

import Button from '../button';

export function Game(props) {
  const { controls, dispatch, editButton, removeButton, game } = props;
  const tags = game.tags.map(tag => (<li key={game.id + '_' + tag.id} id={game.id + '_' + tag.id}>{tag.name}</li>));
  return (
    <li className="game" id={game.id}>
      <h3 className="game__title">{game.title}</h3>
      <ul className="game__details">
        <li><strong>Players:</strong> {game.players.min} - {game.players.max}</li>
        <li>
          <strong>Tags</strong>
          <ul className="game__tag-list">
            {tags}
          </ul>
        </li>
      </ul>
      {controls && (
        <section className="game__buttons">
          <Link
            className={editButton ? '' : 'game__buttons--hidden'}
            to={`/games/${game.id}/edit`}
          >
            <Button game secondary label="Edit" />
          </Link>
          <Button
            className={removeButton ? "btn--remove" : "btn--add"}
            game
            icon={removeButton ? "-" : "+"}
            primary
            onClick={() => dispatch(removeButton ? removeGameFromShelf(game) : addGameToShelf(game))}
            label={removeButton ? "Remove" : "Add"}
          />
        </section>
      )}
    </li>
  );
};

Game.defaultProps = {
  controls: false,
  editButton: true,
  removeButton: true,
  game: {},
};

export default connect()(Game);