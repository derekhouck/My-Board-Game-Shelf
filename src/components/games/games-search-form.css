.games__search {
  text-align: center;
}

.games__search-fields legend {
  text-transform: uppercase;
}

.games__search-buttons > * {
  margin: 1em;
}

@media only screen and (min-width: 768px) {
  .games__search-fields {
    display: grid;
    grid-template-areas: 
      "title input"
      "tags tags";
    justify-content: center;
  }

  .games__search-fields .form-input {
    margin-right: 1em;
  }

  .games__search-title {
    grid-area: title;
  }
  
  .games__search-tags {
    grid-area: tags;
  }
}

@media only screen  and (min-width : 1224px) {
  .games__search-fields {
    display: block;
  }

  .games__search-fields .form-input {
    margin: 0.5em;
  }

  .games__search-buttons > * {
    display: block;
    margin: 1em auto;
  }
}