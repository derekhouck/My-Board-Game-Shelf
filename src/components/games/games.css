.games-header {
  text-align: center;
}

.games h2 {
  text-align: center;
}

.games__no-games {
  text-align: center;
}

.games__list {
  list-style: none;
  margin: 0;
  padding: 0;
}

.game {
  background-color: #4f2119;
  border: 2px solid #19000b;
  box-shadow: 0 1px 4px hsla(9, 52%, 10%, 0.75);
  color: #f9fafa;
  margin: 1em 0;
  padding: 1em 1em 0;
  transition: 0.4s;
}

.game:active, .game:focus, .game:hover {
  box-shadow: 0 12px 16px 0 rgba(0,0,0,0.24), 0 17px 50px 0 rgba(0,0,0,0.19);
}

.game strong {
  font-variant: small-caps;
  font-weight: 400;
}

.game__title {
  border-bottom: 2px solid #f9fafa;
  font-size: 2em;
  font-variant: small-caps;
  font-weight: 400;
  margin: 0 0 0.25em;
  padding-bottom: 0.25em;
  text-align: center;
}

.game__details {
  list-style: none;
  margin: 0;
  padding: 0;
  text-align: center;
}

.game__details li {
  margin: 0.5em 0;
}

.game__tag-list {
  margin: 0;
  padding: 0;
}

.game__tag-list li {
  border: 1px solid #d19cb8;
  color: #d19cb8;
  display: inline-block;
  margin: 0.25em;
  padding: 0.5em;
}

.game__buttons {
  display: flex;
  justify-content: space-between;
  margin-top: 1rem;
}

.game__buttons--hidden {
  visibility: hidden;
}

@media only screen and (min-width: 768px) {
  .games__list {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-around;
  }

  .game {
    display: grid;
    flex: 0 1 45%;
    grid-template-rows: auto 1fr auto;
  }
}

@media only screen  and (min-width : 1224px) {
  .games__wrapper {
    display:grid;
    grid-template-columns: 1fr auto;
    grid-template-areas: 
    "children children"
    "games search-form";
  }
  .games-header {
    grid-area: search-form;
  }

  .games-header > * {
    margin: 0 1em;
  }

  .games__games-list {
    grid-area: games;
  }

  .game {
    flex: 0 1 30%;
    margin: 1em;
    max-width: 25em;
  }
}