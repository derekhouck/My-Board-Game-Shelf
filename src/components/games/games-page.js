import React from 'react';
import { connect } from 'react-redux';
import { fetchGames } from '../../actions/games';
import { fetchUserGames } from '../../actions/users';
import Games from './games';

export class GamesPage extends React.Component {
  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchGames());
  }

  componentDidUpdate(prevProps) {
    const { currentUser, dispatch } = this.props;
    if (!prevProps.loggedIn && this.props.loggedIn) {
      dispatch(fetchUserGames(currentUser.id))
    }
  }

  render() {
    const { games, loggedIn, userGames } = this.props;
    return (
      <section className="games-page">
        <h1>Games</h1>
        <Games
          controls={loggedIn}
          editButton={false}
          games={games}
          userGames={userGames}
        />
      </section>
    );
  }
};

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser,
  games: state.games.games,
  loggedIn: state.auth.currentUser !== null,
  userGames: state.users.games,
});

export default connect(mapStateToProps)(GamesPage);