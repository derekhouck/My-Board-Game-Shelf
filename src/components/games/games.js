import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { fetchUserGames } from '../../actions/users';
import './games.css';

import Button from '../button';
import GamesSearchFrom from './games-search-form';
import Game from './game';

export class Games extends React.Component {
  componentDidMount() {
    const { currentUser, dispatch } = this.props;
    dispatch(fetchUserGames(currentUser.id));
  }

  filterGames(games = []) {
    if (games.length === 0) { return games }
    const { filters } = this.props;
    const re = new RegExp(filters.title, 'i');
    const titleTest = game => filters.title ? re.test(game.title) : true;
    const playersTest = game =>
      filters.players ? (game.players.min <= filters.players && game.players.max >= filters.players) : true;
    const tagIdTest = game =>
      !filters.tagId ? true : filters.tagId === 'null' ? true : (game.tags.filter(tag => tag.id === filters.tagId).length > 0);

    return games.filter(game => titleTest(game) && playersTest(game) && tagIdTest(game));
  }

  renderGames() {
    let body;

    if (this.props.error) {
      body = (
        <div className="message message-error">{this.props.error.message}</div>
      );
    } else if (this.props.loading) {
      body = (
        <div className="message message-default">Loading your games...</div>
      );
    } else {
      let games;
      const filteredGames = this.filterGames(this.props.games);
      if (filteredGames.length === 0) {
        games = (
          <div className="games__no-games">
            <p>You don't have any games at the moment. Let's add one now.</p>
            <Link to="/games/add">
              <Button primary label="Add a game" />
            </Link>
          </div>
        );
      } else {
        games = filteredGames.map(game => <Game key={game.id} game={game} />);
      }
      body = (
        <section className="games">
          <h2>Your Games</h2>
          <ul className="games__list">
            {games}
          </ul>
        </section>
      );
    }
    return body;
  }

  render() {
    return (
      <section className="games__wrapper">
        <header className="games-header">
          <Link to="/games/add" className="centered"><Button primary label="Add a game" /></Link>
          <GamesSearchFrom />
        </header>
        {this.renderGames()}
      </section>
    );
  }
}

const mapStateToProps = state => ({
  currentUser: state.auth.currentUser,
  error: state.users.error,
  games: state.users.games,
  filters: state.games.filters,
  loading: state.loading.loading,
});

export default connect(mapStateToProps)(Games);