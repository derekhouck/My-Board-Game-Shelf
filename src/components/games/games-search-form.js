import React from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, focus } from 'redux-form';
import { filterGames, resetFilters } from '../../actions/games';
import './games-search-form.css';

import Button from '../atoms/button';
import Input from '../atoms/input';
import Select from '../select';
import { separateTags } from '../helpers';

export class GamesSearchForm extends React.Component {
  state = {
    mechanics: [],
    themes: []
  };

  componentDidUpdate(prevProps) {
    const { tags } = this.props;
    if (prevProps.tags !== tags) {
      const tagsObj = separateTags(tags);
      this.setState({
        mechanics: tagsObj.mechanics,
        themes: tagsObj.themes,
      })
    }
  }

  componentWillUnmount() {
    const { dispatch } = this.props;
    return dispatch(resetFilters());
  }
  onSubmit(values) {
    const { mechanics, players, searchTerm, themes } = values;
    const filters = {
      title: searchTerm,
      mechanics,
      players,
      tagId: mechanics && mechanics !== 'null' ? mechanics : themes,
      themes,
    };
    return this.props.dispatch(filterGames(filters));
  }

  render() {
    const { dispatch, handleSubmit, initialValues } = this.props;
    const { mechanics, themes } = this.state;

    return (
      <section className="games__search">
        <form
          className="games__search-form"
          onSubmit={handleSubmit(values => this.onSubmit(values))}
        >
          <fieldset className="games__search-fields">
            <legend>Filter your games</legend>
            <Field
              component={Input}
              name="searchTerm"
              id="searchTerm"
              label="Game title"
              className="games__search-title"
              placeholder={initialValues.title}
              type="text"
            />
            <Field
              component={Input}
              name="players"
              id="players"
              label="Number of players"
              placeholder={initialValues.players}
              type="number"
            />
            <div className="form-input games__search-mechanics">
              <label htmlFor="mechanics">Mechanics</label>
              <Field
                component={Select}
                id="mechanics"
                initialValue={initialValues.mechanics}
                name="mechanics"
                noSelectionLabel='Select a mechanic'
                options={
                  mechanics.map(mechanic => ({ value: mechanic.id, text: mechanic.name }))
                }
              />
            </div>
            <div className="form-input games__search-themes">
              <label htmlFor="themes">Themes</label>
              <Field
                component={Select}
                id="themes"
                initialValue={initialValues.themes}
                name="themes"
                noSelectionLabel='Select a theme'
                options={
                  themes.map(theme => ({ value: theme.id, text: theme.name }))
                }
              />
            </div>
          </fieldset>
          <div className="games__search-buttons">
            <Button label="Search" type="submit" />
            <Button secondary label="Clear filters" type="reset" onClick={() => dispatch(resetFilters())} />
          </div>
        </form>
      </section>
    );
  }
}

GamesSearchForm.defaultProps = {
  initialValues: {}
};

function getUnique(arr, comp) {
  const unique = arr
    .map(e => e[comp])
    // store the keys of the unique objects
    .map((e, i, final) => final.indexOf(e) === i && i)
    // eliminate the dead keys & store unique objects
    .filter(e => arr[e]).map(e => arr[e]);
  return unique;
}

const getTags = (games) => {
  const tagArrays = games.map(game => game.tags);
  const mergedArray = [].concat.apply([], tagArrays);
  const tags = getUnique(mergedArray, 'id');
  return tags;
};

const mapStateToProps = (state, ownProps) => {
  const { games } = ownProps;
  return {
    initialValues: state.games.filters,
    tags: getTags(games)
  }
};

export default connect(mapStateToProps)(
  reduxForm({
    enableReinitialize: true,
    form: 'games-search-form',
    onSubmitFail: (errors, dispatch) => dispatch(focus('games-search-form', Object.keys(errors)[0]))
  })(GamesSearchForm)
);