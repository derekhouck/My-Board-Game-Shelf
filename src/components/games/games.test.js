import React from 'react';
import { shallow } from 'enzyme';
import { Games } from './games';

describe('<Games />', function () {
  it('renders without crashing', () => {
    const callback = jest.fn();
    const wrapper = shallow(<Games currentUser={{}} filters={{}} dispatch={callback} />);
    expect(wrapper.find('.games__wrapper')).toHaveLength(1);
  });
});
