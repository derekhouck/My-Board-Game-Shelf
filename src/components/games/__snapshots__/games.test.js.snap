// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Games /> renders without crashing 1`] = `ShallowWrapper {}`;
